### PR TITLE
Intent and Navigator Rearchitecture Using PrimitiveSetMsg

### PR DESCRIPTION
**Pull request by jonathanlew**
### Description
* replaced ConstPrimitiveVectorPtr with PrimitiveSetMsg outside of handheld controller
* restructured intents and navigator to make adding more navigating primitives easier
  * intents wrap PrimitiveMsg for DirectPrimitiveIntent for simplicity
  * created NavigatorIntent so navigator no longer specifically targets MoveIntents and can generalize

### Testing Done
Existing navigator tests continue to pass
ActionTests needed to be commented out, intend to fix in another PR after new Primitives protos are integrated

### Resolved Issues
part of https://github.com/UBC-Thunderbots/Software/issues/1638

### Review Checklist

**_It is the reviewers responsibility to also make sure every item here has been covered_**

- [ ] **Function & Class comments**: All function definitions (usually in the `.h` file) should have a javadoc style comment at the start of them. For examples, see the functions defined in `thunderbots/software/geom`. Similarly, all classes should have an associated Javadoc comment explaining the purpose of the class.
- [ ] **Remove all commented out code**
- [ ] **Remove extra print statements**: for example, those just used for testing
- [ ] **Resolve all TODO's**: All `TODO` (or similar) statements should either be completed or associated with a github issue
- [ ] **Justify drops in code coverage**: If this PR results in a non-trivial drop in code coverage (a bot should post a coverage diagram as a comment), please justify why we can't test the code that's not covered.

### CODE REVIEW

File: src/software/simulated_tests/simulated_test_simulator.h
Comment (akhilveeraghanta) [inline src/software/simulated_tests/simulated_test_simulator.h]: I understand the problem this is solving, but don't think the name conveys this, its almost misleading

I don't have a better suggestion, but maybe we can come up with something w/ `Wrapper` or `Adapter` (and maybe `Nanopb`, but maybe not)

File: src/software/simulated_tests/simulated_test_simulator.cpp
Comment (akhilveeraghanta) [inline src/software/simulated_tests/simulated_test_simulator.cpp]: Is this our only option to achieve this?

If so, we should expand on the comment above on why we need to hide the proto primitive definition and why its only applicable to simulated tests, and just a bit more justification defending this approach

**Review by akhilveeraghanta — CHANGES_REQUESTED** [review]
Nice work as always 🔥 

after this goes through review, we need to remember to update any outdated documentation as a result of this PR. 

also generally why are we keeping the `Primitive` classes around? the biggest win would be if we can get rid of that layer altogether, if we keep it around it doesn't really make sense to have both imo

**Comment by jonathanlew** [comment]
> also generally why are we keeping the `Primitive` classes around? the biggest win would be if we can get rid of that layer altogether, if we keep it around it doesn't really make sense to have both imo

This PR is working with the old PrimitiveMsg, so I wanted to use the existing ProtoCreatorVisitor instead of making new constructors. The issue for the new constructors for the new PrimitiveMsgs is here: https://github.com/UBC-Thunderbots/Software/issues/1639

File: src/software/ai/navigator/navigator.cpp
Comment (jonathanlew) [reply src/software/ai/navigator/navigator.cpp]: basically navigator is checking the output of the path manager, but I'll add a comment that this warning came from the navigator

File: src/software/ai/ai.h
Comment (garethellis0) [inline src/software/ai/ai.h:28]: I think `Primitives` makes more sense then the type here since this is supposed to be a high level semantic explanation?

File: src/software/ai/intent/chip_intent.cpp
Comment (jonathanlew) [reply src/software/ai/intent/chip_intent.cpp]: https://github.com/UBC-Thunderbots/Software/issues/1639 will use the new protos defined in https://github.com/UBC-Thunderbots/Software/pull/1620

File: src/software/ai/navigator/navigator.cpp
Comment (jonathanlew) [reply src/software/ai/navigator/navigator.cpp]: nuked

File: src/software/simulated_tests/simulated_test_simulator.h
Comment (jonathanlew) [reply src/software/simulated_tests/simulated_test_simulator.h]: changed to SerializedToNanoPbSimulatorAdapter

File: src/software/simulated_tests/simulated_test_simulator.cpp
Comment (jonathanlew) [reply src/software/simulated_tests/simulated_test_simulator.cpp]: actually, I can forward Simulator and hold shared_ptr to it

File: src/software/ai/intent/intent.h
Comment (jonathanlew) [reply src/software/ai/intent/intent.h]: :( I really think we can assume meters per second in Software

File: src/software/ai/intent/intent.h
Comment (jonathanlew) [reply src/software/ai/intent/intent.h]: fixed

File: src/shared/proto/tbots_software_msgs.proto
Comment (jonathanlew) [reply src/shared/proto/tbots_software_msgs.proto]: Turns out I don't need it anymore, so reverted

File: src/software/ai/hl/stp/action/chip_action_test.cpp
Comment (jonathanlew) [reply src/software/ai/hl/stp/action/chip_action_test.cpp:10]: added TODO

File: src/software/ai/intent/intent.h
Comment (jonathanlew) [reply src/software/ai/intent/intent.h]: hm, this get overridden by every intent that wants to navigate. I'm not sure how we can integrate this into navigator

Comment (akhilveeraghanta) [reply src/software/ai/intent/intent.h]: Sorry I wasn't clear here, wasn't saying we should integrate this into the Navigator

We should limit the scope of this function to just the navigator by making it protected, and then making the Navigator a friend of this class, so that it can access that function.

Its a really good example of when you should use `friend` imo, the primitive msg tracked inside is internal to the intent, and only the navigator cares about updating it.

**Comment by jonathanlew** [comment]
@MathewMacDougall requesting a MatCI:tm:

**Comment by MathewMacDougall** [comment]
Current MatCI results

Communicating with robots but can't move wheels. The robot is charging / discharging fine, just not moving.

**Comment by MathewMacDougall** [comment]
Ah just didn't have the right firmware. MatCI :tm: passed :heavy_check_mark: 

Validated we can give the RadioBackend a PrimitiveSetMsg and it will be properly sent to the robots.

**Comment by MathewMacDougall** [comment]
> also generally why are we keeping the Primitive classes around? the biggest win would be if we can get rid of that layer altogether, if we keep it around it doesn't really make sense to have both imo

@akhilveeraghanta I'm assuming you mean removing the `Primitive` classes in favor or the proto versions right? That seems fine if we're ok with having proto "deeper" in the navigator as we do currently, otherwise would be a way to abstract proto out of stuff. Definitely a strong argument for not having both, so up to you and @jonathanlew

File: src/software/ai/intent/intent.h
Comment (jonathanlew) [reply src/software/ai/intent/intent.h]: Yeah that's fair, so then we want intent test classes be friends of Intent too?

**Review by jordanzliu — CHANGES_REQUESTED** [review]

File: src/software/ai/intent/intent.h
Comment (jonathanlew) [reply src/software/ai/intent/intent.h:14]: made https://github.com/UBC-Thunderbots/Software/issues/1651 as making this change is out of scope, since BallCollisionType is used in Actions when constructing Intents

File: src/software/ai/intent/intent.cpp
Comment (MathewMacDougall) [inline src/software/ai/intent/intent.cpp]: I'm curious what a case would be that these navigator params can't be provided / derived directly from the primitive params? ie. is there an example of how we would use these aside form the MoveIntent?

File: src/software/ai/intent/movespin_intent.h
Comment (MathewMacDougall) [inline src/software/ai/intent/movespin_intent.h]: What's the difference between this and a `SpinningMoveIntent`?

File: src/software/simulated_tests/serialized_to_nano_pb_simulator_adapter.h
Comment (MathewMacDougall) [inline src/software/simulated_tests/serialized_to_nano_pb_simulator_adapter.h]: Can you clarify what the purpose of this wrapper class is? I see it's to prevent uses of the class from having to depend on NanoPb protobufs, but why? Seems like it's to maintain the nicer interface of providing "primitives" rather than bytes, because we can't have both the nanopb and google protobuf headers together.

I think we could avoid this with the conversion code @garethellis0 wrote (I don't see it in `master`, is it still part of your trajectory planner branch?). Then callers are free to provide / create google protobufs, they just need to call a conversion function to make it a nanopb protobuf before giving it to the simulator. Then we don't have to add another layer just to really abstract 1-2 functions.

**Review by MathewMacDougall — CHANGES_REQUESTED** [review]
@jonathanlew can you please elaborate on how you see the Intent `getNavigatorParams` working?

Right now it just seems like they're used to identify if the navigator should do anything with that Intent (ie. navigate), and the other times it actually uses the data it seems to mostly just be used in place of the data for a MoveIntent. Could you please give an example of how this might be used / useful for something that's not a `MoveIntent`? Right now I'm having a hard time understanding the benefits.

File: src/software/ai/intent/movespin_intent.h
Comment (jonathanlew) [reply src/software/ai/intent/movespin_intent.h]: whoops bad merge, movespin should not exist

File: src/software/simulated_tests/serialized_to_nano_pb_simulator_adapter.h
Comment (jonathanlew) [reply src/software/simulated_tests/serialized_to_nano_pb_simulator_adapter.h]: Since we cannot include google protobuf and nanopb in the same target (without some unsavoury namespacing), we cannot have a function that directly converts google protobuf to nanopb. @garethellis0 's solution was to convert `Primitive` to nanopb, bypassing google protobuf entirely. Since we've decided to delete the `Primitive` classes and use protobug `PrimitiveMsg`s directly, we need some way to avoid including google protobufs and nanopb in the same target, and this Adapter seemed like the only way

File: src/software/ai/intent/intent.cpp
Comment (jonathanlew) [reply src/software/ai/intent/intent.cpp]: Eventually, we will have `MoveIntent`, `AutokickIntent`, and `AutochipIntent`, each with their namesake PrimitiveMsg internally. Each will all want to use path planning and will override NavigatorParams. However, I think we should decouple "having a particular `PrimitiveMsg`" from "I want to use the path planner". First, I don't think the Navigator should digging around PrimitiveMsgs. Second, I think PivotIntent, for example, might define a small trajectory for a MovePrimitiveMsg that it doesn't want the navigator to interfere with.

Comment (jonathanlew) [reply src/software/ai/intent/intent.cpp]: I'll try to come up with a better way to do this

File: src/software/ai/hl/stp/action/intercept_ball_action_test.cpp
Comment (garethellis0) [inline src/software/ai/hl/stp/action/intercept_ball_action_test.cpp:12]: Sorry if this is explained elsewhere, why are these tests commented out? Do we still have coverage on these actions?

File: src/software/ai/intent/intent.h
Comment (garethellis0) [inline src/software/ai/intent/intent.h]: I'm a little confused about how we're using this function. Could you explain at a high level why it's needed?

File: src/software/ai/navigator/navigator.cpp
Comment (garethellis0) [inline src/software/ai/navigator/navigator.cpp]: It would make sense to have a `getPathObjectiveFromIntent` function since there's nothing shared between different iterations of this loops except for `path_objectives`

File: src/software/simulated_tests/serialized_to_nano_pb_simulator_adapter.h
Comment (garethellis0) [reply src/software/simulated_tests/serialized_to_nano_pb_simulator_adapter.h]: Jon's correct, the adapter I wrote won't work here. Given that the only time we provide actual NanoPb messages to the simulator is during tests, why not change the simulator interface to only accept serialized primitives and add a helper function for the tests?

**Review by garethellis0 — CHANGES_REQUESTED** [review]

File: src/software/ai/hl/stp/action/intercept_ball_action_test.cpp
Comment (jonathanlew) [reply src/software/ai/hl/stp/action/intercept_ball_action_test.cpp:12]: This PR makes *Intent not extend *Primitive, so the functions like getDestination() don't work anymore. However, after this PR we're still using PrimitiveParamMsg, so to make these tests work, I'd have to change all the tests to look at individual params. After #1643, I can reenable the tests by accessing the fields in the new proto, which are much more user friendly

File: src/software/ai/intent/intent.h
Comment (jonathanlew) [reply src/software/ai/intent/intent.h]: Example: MoveIntent starts with a PrimitiveMsg to go to point C, starting from point A. Navigator finds a path that goes from A to B to C to avoid an obstacle. As MovePrimitives are currently, this means that the robot should go to point B and MovePrimitiveMsg needs to be changed to go to point B. 

getUpdatedPrimitiveMsg takes the new destination and final speed derived from the path, updates the PrimitiveMsg and returns the updated PrimitiveMsg

File: src/software/ai/ai.h
Comment (jonathanlew) [reply src/software/ai/ai.h]: fixed

File: src/software/ai/navigator/navigator.cpp
Comment (jonathanlew) [reply src/software/ai/navigator/navigator.cpp]: refactored, but in a different way

File: src/software/ai/intent/intent.h
Comment (jonathanlew) [reply src/software/ai/intent/intent.h]: yeah I don't mind either way

File: src/software/ai/intent/intent.h
Comment (garethellis0) [reply src/software/ai/intent/intent.h]: This feels a bit hacky, like we're moving some of the navigator work into each primitive. 

Based on this, why not keep some sort of Intent->PrimitiveMsg visitor that takes this information + an intent? This would allow decoupling of intents from primitives entirely, which seems like a better option from a architecture perspective.

Comment (jonathanlew) [reply src/software/ai/intent/intent.h]: A visitor would abstract away PrimitiveMsg, but it would require each Intent to store all the fields necessary to create a PrimitiveMsg, which is quite a bit of duplication. To add a new primitive, in addition to implementing in firmware and defining in protobuf, one would have to define it again as an Intent. 

I'll poke around with visitor on PrimitiveMsg to update them with the Path. I think that would be simplest

File: src/software/ai/intent/intent.cpp
Comment (MathewMacDougall) [reply src/software/ai/intent/intent.cpp]: Ok yea I get what you're going for. Happy to help brainstorm solutions, there's definiltey a few points that could introduce confusion (ie. if an intent is navigating, how does one know what the navigator is responsible for vs the primitive)

File: src/software/simulated_tests/serialized_to_nano_pb_simulator_adapter.h
Comment (MathewMacDougall) [reply src/software/simulated_tests/serialized_to_nano_pb_simulator_adapter.h]: @garethellis0 doesn't the adapter convert Google Protobufs to NanoPb?

If the conversion doesn't work though I'd prefer Gareth's suggestion over adding a whole new layer of abstraction since the Simulator is already pretty deep in terms of composed layers (sim, threaded sim, standalone sim)

Comment (jonathanlew) [reply src/software/simulated_tests/serialized_to_nano_pb_simulator_adapter.h]: it's not possible to write a converter from Google Protobufs to NanoPb since we can't include two definitions of the same name. `ThreadedSimulator` uses the NanoPb Protobuf interface of `Simulator` so we can't easily remove that. I don't see how to do this

Comment (garethellis0) [reply src/software/simulated_tests/serialized_to_nano_pb_simulator_adapter.h]: Yeah, Jon is correct here. I think the simplest solution is to just use only `std::vector<uin8_t>` (ie. a serialized mesage) and then have a helper functions in for tests.

Comment (jonathanlew) [reply src/software/simulated_tests/serialized_to_nano_pb_simulator_adapter.h]: `serialized_to_nano_pb_simulator_adapter` is a `test_only` target that lived in the `simulated_test` folder and is only used by the `SimulatedTestFixture`. I think this is better than changing the entire `StandaloneSimulator`->`ThreadedSimulator`->`Simulator` stack to only have `std::vector<uint8_t>` in their interfaces

Comment (MathewMacDougall) [reply src/software/simulated_tests/serialized_to_nano_pb_simulator_adapter.h]: `serialized_to_nano_pb_simulator_adapter` still adds _another_ layer to the simulator though that will also have to be updated everytime functionality is added to the interface (eg. the upcoming changes to add/remove robots), which just makes it more of a pain to support IMO since we already have several layers of composition.

@jonathanlew is your main issue with changing the stack to use `std::vector<uint8_t>` interfaces due to the work of actually making the change, or because you feel it's a worse interface?

Comment (jonathanlew) [reply src/software/simulated_tests/serialized_to_nano_pb_simulator_adapter.h]: `serialized_to_nano_pb_simulator_adapter` only needs to change if the `SimulatedTestFixture` requires more functionality to be exposed, not every time the `Simulator` gets more functionality. 

Yes, it requires a lot of work to make it work, but I def take issue with changing the entire stack to accommodate a quirk of the `SimulatedTestFixture`. Changing the stack to use `std::vector<uint8_t>` is a worse interface because it obfuscates the type that it represents and makes it possible to fail in more places. As it is currently, when running with the `StandaloneSimulator` binary, deserializing to nanopb proto can only fail in the nanopb listener. If we instead listened for `std::vector<uint8_t>` and passed that along `StandaloneSimulator`->`ThreadedSimulator`->`Simulator`, the failure can happen anywhere along that stack if we have to deserialize for whatever reason. Not to mention `StandaloneSimulatorTest`s, `ThreadedSimulatorTest`, and `SimulatorTest`s will have the overhead of serializing all primitives before passing them in

Comment (MathewMacDougall) [reply src/software/simulated_tests/serialized_to_nano_pb_simulator_adapter.h]: Ok fair enough, I won't block this PR on this comment. Once this is merged and I finish adding more sim functionality we can see what it looks like.

File: src/software/ai/intent/intent.cpp
Comment (jonathanlew) [reply src/software/ai/intent/intent.cpp]: I've tried to decouple it some more by cutting back on the intent abstraction. Please take another look

File: src/software/ai/intent/chip_intent.h
Comment (MathewMacDougall) [inline src/software/ai/intent/chip_intent.h]: Not sure how I feel about only some `Intents` hacing a `generatePrimitive` function. This is not consistent between all primitives, and to me makes it unclear what code is responsible for generating primitives. The Intents, or the Navigator.

File: src/software/ai/navigator/navigator.cpp
Comment (MathewMacDougall) [inline src/software/ai/navigator/navigator.cpp]: Possibly confusing that even "non-navigating intents" just generate degenerate paths that the visitor then ignores.

**Review by MathewMacDougall — CHANGES_REQUESTED** [review]
Keeping review pretty high-level again and basically just looked at the Intents and Navigator.

I'd still really like to see a concrete example or 2 of how optionally providing a "navigation destination" for various Intents would be used. Probably worth a call on Saturday.

**Comment by jonathanlew** [comment]
> 
> 
> Keeping review pretty high-level again and basically just looked at the Intents and Navigator.
> 
> I'd still really like to see a concrete example or 2 of how optionally providing a "navigation destination" for various Intents would be used. Probably worth a call on Saturday.

MoveIntent provides those parameters

File: src/software/simulated_tests/serialized_to_nano_pb_simulator_adapter.h
Comment (MathewMacDougall) [reply src/software/simulated_tests/serialized_to_nano_pb_simulator_adapter.h]: @jonathanlew Do you want to wait for #1662 to merge before this PR so we can remove the adapter, or put this through first?

Comment (jonathanlew) [reply src/software/simulated_tests/serialized_to_nano_pb_simulator_adapter.h]: I'm gonna remove the adapter

**Review by HaomiaoZ — COMMENTED** [review]
Overall looks good to me. Just have couple nits

File: src/software/ai/navigator/navigating_primitive_creator.cpp
Comment (jonathanlew) [reply src/software/ai/navigator/navigating_primitive_creator.cpp:30]: hm, rather not put emojis in the code base haha

File: src/software/ai/intent/move_intent.h
Comment (jonathanlew) [reply src/software/ai/intent/move_intent.h:50]: fixed comments above

**Review by MathewMacDougall — DISMISSED** [review]

**Comment by jonathanlew** [comment]
@garethellis0 @jordanzliu @akhilveeraghanta I need re-review since you requested changes

**Review by akhilveeraghanta — DISMISSED** [review]
The proto package really made the difference, nice work 🎉

File: src/software/ai/intent/intent.h
Comment (garethellis0) [inline src/software/ai/intent/intent.h:116]: Quick comment like `The id of the robot this intent is to be run on` would be helpful here.

File: src/software/ai/intent/intent_visitor.h
Comment (garethellis0) [inline src/software/ai/intent/intent_visitor.h:38]: Why don't we allow visiting of the other intents here?

**Review by garethellis0 — CHANGES_REQUESTED** [review]
Generally this looks good! I think the navigator-intent-primitive interactions are much better now.

File: src/software/ai/intent/intent_visitor.h
Comment (jonathanlew) [reply src/software/ai/intent/intent_visitor.h:38]: The subclasses of `DirectPrimitiveIntent` are completely covered by visiting `DirectPrimitiveIntent`

File: src/software/ai/intent/intent.h
Comment (jonathanlew) [reply src/software/ai/intent/intent.h:116]: added

File: src/software/ai/intent/intent_visitor.h
Comment (garethellis0) [reply src/software/ai/intent/intent_visitor.h:38]: Could we leave a quick comment indicating that here?

Comment (jonathanlew) [reply src/software/ai/intent/intent_visitor.h:38]: added comments

File: src/software/ai/navigator/navigating_primitive_creator.h
Comment (jordanzliu) [inline src/software/ai/navigator/navigating_primitive_creator.h:34]: I was curious about the rationale for having `NavigatingPrimitiveCreator` responsible for velocity around path knots and enemy robots, instead of having the `Navigator` responsible for it and velocity fully encapsulated in `Path`.

**Review by jordanzliu — CHANGES_REQUESTED** [review]
LGTM other than the question I had

File: src/software/ai/navigator/navigating_primitive_creator.h
Comment (jonathanlew) [reply src/software/ai/navigator/navigating_primitive_creator.h:34]: The idea is that (NavigatingIntent, Path) in and Primitive out should support trajectory planner as well, i.e. Path and NavigatingIntent in and trajectory planner primitive out, so that's why I chose that abstraction. @garethellis0 thoughts?

Comment (jonathanlew) [reply src/software/ai/navigator/navigating_primitive_creator.h:34]: essentially, this is a _temporary_ solution before trajectory planner integration. With the trajectory planner, we'll be sending the entire path to the robots, not just one final speed and one final destination

**Comment by jonathanlew** [comment]
@garethellis0 @jordanzliu I think I've addressed your comments can I please get reviews?

File: src/software/ai/navigator/navigating_primitive_creator.h
Comment (jordanzliu) [reply src/software/ai/navigator/navigating_primitive_creator.h:34]: The way I understand this is, both `PathManager` and `NavigatingPrimitiveCreator` will both be responsible for some sort of obstacle avoidance functionality?

Comment (jonathanlew) [reply src/software/ai/navigator/navigating_primitive_creator.h:34]: That's kind of right. `PathManager` creates paths that don't intersect with each other and don't intersect obstacles, but it can't control the speed of the robots, because `Path` can't express speed as it is now. The `NavigatingPrimitiveCreator` directly translates these paths into primitives so that the robot can follow the path, but it additionally slows robots down near enemy robot obstacles to be more safe. It's not super ideal as it is and we're expecting some changes to this with trajectory planner as well (right after primitives are integrated), so we can think of better ways to do this. @garethellis0 has some ideas with a spline navigator and integrating the trajectory planner into path planning that should improve this as well

Comment (jordanzliu) [reply src/software/ai/navigator/navigating_primitive_creator.h:34]: Ok, I see now, looks like there's a plan to deal with this down the road. LGTM

**Review by jordanzliu — APPROVED** [review]
💯

**Review by MathewMacDougall — APPROVED** [review]

**Review by akhilveeraghanta — APPROVED** [review]

**Review by garethellis0 — APPROVED** [review]
LGTM :+1: :tada:

**Pull request merged**
